vochain/apputils.go

```diff
@@ -29,102 +28,118 @@ import (
-			caPubk, err = ethereum.PubKeyFromSignature(caBundle, p.GetSignature())
-			if err != nil {
-				return false, nil, fmt.Errorf("cannot fetch ca address from signature: %w", err)
-			}
-			if !bytes.Equal(caPubk, censusRoot) {
-				return false, nil, fmt.Errorf("ca bundle signature does not match")
-			}
-		case models.ProofCA_ECDSA_BLIND:
-			// Blind CA check
-			pubdesc, err := ethereum.DecompressPubKey(censusRoot)
-			if err != nil {
-				return false, nil, fmt.Errorf("cannot decompress CA public key: %w", err)
-			}
-			pub, err := blind.NewPublicKeyFromECDSA(pubdesc)
-			if err != nil {
-				return false, nil, fmt.Errorf("cannot compute blind CA public key: %w", err)
-			}
-			signature, err := blind.NewSignatureFromBytes(p.GetSignature())
-			if err != nil {
-				return false, nil, fmt.Errorf("cannot compute blind CA signature: %w", err)
-			}
-			if !blind.Verify(new(big.Int).SetBytes(ethereum.HashRaw(caBundle)), signature, pub) {
-				return false, nil, fmt.Errorf("blind CA verification failed %s", log.FormatProto(p.Bundle))
-			}
-		default:
-			return false, nil, fmt.Errorf("ca proof %s type not supported", p.Type.String())
-		}
-		return true, big.NewInt(1), nil
-
-	case models.CensusOrigin_ERC20:
-		p := proof.GetEthereumStorage()
-		if p == nil {
-			return false, nil, fmt.Errorf("ethereum proof is empty")
+		pub, err := blind.NewPublicKeyFromECDSA(pubdesc)
+		if err != nil {
+			return false, nil, fmt.Errorf("cannot compute blind CA public key: %w", err)
 		}
-		if !bytes.Equal(p.Key, key) {
-			return false, nil, fmt.Errorf("proof key and leafData do not match (%x != %x)", p.Key, key)
+		signature, err := blind.NewSignatureFromBytes(p.GetSignature())
+		if err != nil {
+			return false, nil, fmt.Errorf("cannot compute blind CA signature: %w", err)
 		}
-		hexproof := []string{}
-		for _, s := range p.Siblings {
-			hexproof = append(hexproof, fmt.Sprintf("%x", s))
+		if !blind.Verify(new(big.Int).SetBytes(ethereum.HashRaw(caBundle)), signature, pub) {
+			return false, nil, fmt.Errorf("blind CA verification failed %s", log.FormatProto(p.Bundle))
 		}
-		amount := big.Int{}
-		amount.SetBytes(p.Value)
-		hexamount := hexutil.Big(amount)
-		log.Debugf("validating erc20 storage proof for key %x and amount %s", p.Key, amount.String())
-		valid, err := ethstorageproof.VerifyEthStorageProof(
-			&ethstorageproof.StorageResult{
-				Key:   fmt.Sprintf("%x", p.Key),
-				Proof: hexproof,
-				Value: &hexamount,
-			},
-			ethcommon.BytesToHash(censusRoot),
-		)
-		return valid, &amount, err
+	default:
+		return false, nil, fmt.Errorf("ca proof %s type not supported", p.Type.String())
 	}
-	return false, nil, fmt.Errorf("proof type not supported for census origin %d", censusOrigin)
+	return true, big.NewInt(1), nil
+}
+
+// VerifyProofERC20 verifies a proof with census origin ERC20 (mapbased).
+// Returns verification result and weight.
+func VerifyProofERC20(process *models.Process, proof *models.Proof, censusOrigin models.CensusOrigin,
```

Comment:
I also wonder if it would make sense to write an interface around this functions, so it will be more modular and easy to extend with new proofs. What do you think?

Reply:
I have introduced the `VerifyProofFunc` function type to define a prototype of the verification function that all verifyProof functions follow.  I haven't followed the interface approach because interfaces work for methods on structs, but in this case we only have standalone functions whose prototype can be defined by a function type.  Let me know what you think about it!

Reply:
Yes, that's a better approach IMHO. Thanks